hack/generator/pkg/astmodel/armconversion/from_arm_conversion_function.go

```diff
@@ -0,0 +1,431 @@
+	inputIdent := ast.NewIdent("armInput")
+	typedInputIdent := ast.NewIdent("typedInput")
+
+	// perform a type assert
+	fromArmStatements = append(fromArmStatements, &ast.AssignStmt{
+		Lhs: []ast.Expr{
+			typedInputIdent,
+			ast.NewIdent("ok"),
+		},
+		Tok: token.DEFINE,
+		Rhs: []ast.Expr{
+			&ast.TypeAssertExpr{
+				X:    inputIdent,
+				Type: ast.NewIdent(c.armTypeName.Name()),
+			},
+		},
+	})
+
+	// Check the result of the type assert
+	fromArmStatements = append(fromArmStatements, &ast.IfStmt{
+		Cond: &ast.UnaryExpr{
+			Op: token.NOT,
+			X:  ast.NewIdent("ok"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				&ast.ReturnStmt{
+					Results: []ast.Expr{
+						&ast.CallExpr{
+							Fun: &ast.SelectorExpr{
+								X:   ast.NewIdent("fmt"),
+								Sel: ast.NewIdent("Errorf"),
+							},
+							Args: []ast.Expr{
+								&ast.BasicLit{
+									Kind:  token.STRING,
+									Value: fmt.Sprintf("\"unexpected type supplied for FromArm function. Expected %s, got %%T\"", c.armTypeName.Name()),
+								},
+								inputIdent,
+							},
+						},
+					},
+				},
+			},
+		},
+	})
+
+	definedErr := false
+
+	fromArmFieldConversionWrapper := func(toProp *astmodel.PropertyDefinition, fromType *astmodel.ObjectType) []ast.Stmt {
+		return c.fromArmFieldConversionHandler(codeGenerationContext, methodName, typedInputIdent, receiverIdent, toProp, fromType, &definedErr)
+	}
+
+	// Do all of the assignments
+	fromArmStatements = append(
+		fromArmStatements,
+		generateTypeConversionAssignments(
+			c.armType,
+			kubeType,
+			fromArmFieldConversionWrapper)...)
+	fromArmStatements = append(fromArmStatements, &ast.ReturnStmt{
+		Results: []ast.Expr{
+			ast.NewIdent("nil"),
+		},
+	})
+
+	fromArmFunc := astbuilder.DefineFunc(
+		astbuilder.FuncDetails{
+			Name:          ast.NewIdent(methodName),
+			ReceiverIdent: receiverIdent,
+			ReceiverType: &ast.StarExpr{
+				X: ast.NewIdent(receiver.Name()),
+			},
+			Comment: "converts from an Azure ARM object to a Kubernetes CRD object",
+			Params: []*ast.Field{
+				{
+					Type: &ast.SelectorExpr{
+						X:   ast.NewIdent(astmodel.GenRuntimePackageName),
+						Sel: ast.NewIdent("KnownResourceReference"),
+					},
+					Names: []*ast.Ident{
+						ast.NewIdent("owner"),
+					},
+				},
+				{
+					Type: ast.NewIdent("interface{}"),
+					Names: []*ast.Ident{
+						inputIdent,
+					},
+				},
+			},
+			Returns: []*ast.Field{
+				{
+					Type: ast.NewIdent("error"),
+				},
+			},
+			Body: fromArmStatements,
+		})
+
+	return fromArmFunc
+}
+
+func (c *ArmConversionFunction) fromArmComplexPropertyConversion(
+	source ast.Expr,
+	destination ast.Expr,
+	destinationType astmodel.Type,
+	methodName string,
+	nameHint string, // TODO: remove this or clarify it?
+	codeGenerationContext *astmodel.CodeGenerationContext,
+	conversionContext []astmodel.Type,
+	assignmentHandler func(result ast.Expr, destination ast.Expr) ast.Stmt) []ast.Stmt {
```

Comment:
Some of these method arg lists are getting quite long... Not a major but might point to some cleaner abstraction.

Reply:
Moved to a builder pattern where static data could reside on the enclosing builder object -- I think they're back down to more reasonable numbers (a _bit_ big but not this bad). Let me know what you think.